File: backend/models/job.py
"""Pydantic models for table generation jobs."""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class JobStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETE = "complete"
    FAILED = "failed"
    CANCELLED = "cancelled"


class JobRecord(BaseModel):
    """One row in the job queue."""
    job_id: str
    study_id: str
    table_id: str                             # shell id, e.g. 't_14_3_1_2'
    table_number: str
    status: JobStatus
    submitted_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    output_path: str | None = None
    error: str | None = None
    triggered_by: str = "user"
    batch_id: str | None = None


class JobSubmitRequest(BaseModel):
    """Body of POST /studies/{id}/jobs.

    Submitting multiple table_ids creates a batch.
    """
    table_ids: list[str]
    triggered_by: str = "user"


class JobSubmitResponse(BaseModel):
    """Response from POST /studies/{id}/jobs."""
    batch_id: str | None
    jobs: list[JobRecord]


class BatchProgress(BaseModel):
    """Aggregated progress for the batch progress bar."""
    batch_id: str
    total: int
    complete: int
    failed: int
    running: int
    queued: int
    started_at: datetime
    completed_at: datetime | None = None

File: frontend/src/types/job.ts
export type JobStatus = "queued" | "running" | "complete" | "failed" | "cancelled";

export interface JobRecord {
  job_id: string;
  study_id: string;
  table_id: string;
  table_number: string;
  status: JobStatus;
  submitted_at: string;
  started_at: string | null;
  completed_at: string | null;
  output_path: string | null;
  error: string | null;
  triggered_by: string;
  batch_id: string | null;
}

export interface JobSubmitResponse {
  batch_id: string | null;
  jobs: JobRecord[];
}

export interface TablePreviewData {
  kind?: "table";
  shell_id: string;
  title: [string, string, string];
  header_text: string;
  column_headers: string[];
  arm_n_labels: string[];
  body_rows: string[][];
  footnotes: { kind: string; text: string }[];
  source: string;
  page_indicator: string;
}

export interface FigurePreviewData {
  kind: "figure";
  shell_id: string;
  title: [string, string, string];
  header_text: string;
  image: string; // data URL (data:image/png;base64,...)
  source: string;
  page_indicator: string;
}

export type PreviewData = TablePreviewData | FigurePreviewData;

// Review workflow: pending (QC outstanding) -> qc_passed | qc_failed ->
// approved (biostat signed off).
export type OutputStatus = "pending" | "qc_passed" | "qc_failed" | "approved";

export interface OutputRecord {
  output_id: string;
  filename: string;
  table_number: string;
  table_id: string;
  population: string;
  generated_at: string;
  size_bytes: number;
  status: OutputStatus;
  audit_path: string | null;
}

export interface QcChecklistItem {
  id: string;
  label: string;
  result: "pass" | "fail" | "na";
  comment: string;
}

export interface QcRecord {
  reviewer: string;
  performed_at: string;
  result: "pass" | "fail";
  items: QcChecklistItem[];
  comments: string;
  auto_checks: Record<string, unknown>;
}

export interface SignoffRecord {
  name: string;
  role: string;
  signed_at: string;
  comment: string;
  qc_reviewer: string;
  qc_performed_at: string;
}

export interface OutputAudit {
  generated?: {
    at: string;
    table_id: string;
    filename: string;
    data_extract_date: string;
  };
  qc?: QcRecord;
  signoff?: SignoffRecord;
  review_history?: Record<string, unknown>[];
}
